Linux packaging

ppisa opened the pull request:
Project switched to CMake and made the initial attempt for proper GNU/Linux packaging.

The Open SUSE Build service builds package for Fedora an SUSE RPMs distributions, but there is some problem to investigate for Debian based distributions. I am not sure if it s not problem of OBS, because reported missing packages are found on Debian and Ubuntu packages lists

https://build.opensuse.org/package/show/home:ppisa/zero-elabviewer

The Ubuntu build attempt is done on respective Launchpad service

https://code.launchpad.net/~cvut-fel-sdi/+recipe/zero-elabviewer-build

Packages can be downloaded from SDI PPA (Personal Package Archive)

https://code.launchpad.net/~cvut-fel-sdi/+archive/ubuntu/ppa

```
sudo add-apt-repository ppa:cvut-fel-sdi/ppa
sudo apt update
sudo apt install zero-elabviewer
```
The main application name `zero_elabviewer` but the package name has been changed to `zero-elabviewer` because underscore is forbidden character in the Debian/Ubuntu package names.

It the frame of of changes required for GNU/Linux packaging, MAC OS X support has been added as well.

The build has not been tested with any real device on GNU/Linux yet.

JakubVanek committed:
fix: Remove problematic color profile from CTU logo files

Currently, Zero eLab Viewer prints the following warning message during
startup:

> libpng warning: iCCP: known incorrect sRGB profile

The CTU logo PNGs contain the IEC61966-2.1 color profile. This profile
reportedly can cause problems [1] and is non-free [2].

To resolve the issue and stop the warning, this commit alters the images
to remove the color profile. The images look the same in GIMP.
The files were updated using the following command [3]:

```
$ convert image.png +profile "icc" image.png
```

[1]: https://github.com/pnggroup/libpng/blob/68d7ce8040a611e9166890e02ecc409a03d76c2e/png.c#L2372
[2]: https://github.com/google/cadvisor/issues/915
[3]: https://github.com/nodejs/node/pull/5811

JakubVanek committed:
fix: Add support for the Qt 6 way of registering QML elements

See https://www.qt.io/blog/qml-type-registration-in-qt-5.15

The way Zero eLab Viewer registered its QML elements is no longer
present in Qt 6.

JakubVanek committed:
feat: Upgrade QCustomPlot to version 2.1.1

This version should be compatible with up to Qt 6.4. Local tests show
that it works even on Qt 6.7.2.

JakubVanek committed:
fix: Use the Qt5.15+ button group id-pressed signal when available

The `buttonPressed(int)` signal is deprecated in Qt 5.15 and
it is removed in Qt 6.

JakubVanek committed:
fix: Do not crash when the app is exited via the menu bar on Qt6

Qt 6 on Wayland calls the closeEvent function from within the qApp->quit()
function. This caused a double free of the PWM dialog.

This commit fixes this by moving the freeing into the main window destructor.

ppisa committed:
Switch to CMake and initial GNU/Linux packaging attempt

Co-authored-by: Jakub Vaněk <linuxtardis@gmail.com>
Signed-off-by: Pavel Pisa <ppisa@pikron.com>

ppisa committed:
Debian package is forbidden to include "_" in the name

Signed-off-by: Pavel Pisa <ppisa@pikron.com>

ppisa committed:
CMake and GNU/Linux packaging corrections

Signed-off-by: Pavel Pisa <pisa@fel.cvut.cz>

JakubVanek committed:
build: Import basic CI config from QtRvSim

JakubVanek committed:
feat: Improve the desktop file

This integrates the changes made in
https://github.com/adamberlinger/zero_elabviewer/pull/3

JakubVanek committed:
chore: Rename APPID not to include FEE CTU in the name

JakubVanek committed:
build: Add Qt 6 back to CI

JakubVanek committed:
fix: Resolve a warning about function exiting without returning a value

JakubVanek committed:
feat: Introduce a new icon

This is the icon designed in
https://github.com/adamberlinger/zero_elabviewer/pull/3 .

JakubVanek committed:
feat: Add udev rule for STM32F042

JakubVanek committed:
fix: Generate the Debian package as version 1.0-2

JakubVanek committed:
feat: Update copyright and lintian-overrides

JakubVanek committed:
feat: Provide a hint for permission error resolution on Linux

Also adds section about adding yourself to a relevant group

ppisa committed:
Rename and change homepage repository link

The step done in hope to unite workforce on single project.

Signed-off-by: Pavel Pisa <pisa@fel.cvut.cz>

JakubVanek commented:
Commit 325a65d6ad38868c2515f01741fc9013f340c407 není úplný - v Debian balíčku se udev pravidlo pořád instaluje. Revert commitu d0aaa6c3aafc876224c281055ef1ed2e4c88d08e bych ale doufal, že kolem udev pravidla odstraní všechno potřebné.

JakubVanek commented:
Ten `uaccess` se mi ale zdál ještě jako relativně omezená míra zásahu (vs. MODE=0666 nebo match na všechny zařízení). Podobné řešení vlastně dělá https://github.com/hroncok/3dprinter-udev-rules (`ID_MAKER_TOOL` nepřímo přidá `uaccess` i v čistém systemd: https://github.com/systemd/systemd/pull/2844) a právě mnou již zmiňovaný https://salsa.debian.org/debian/avrdude/-/blob/master/debian/avrdude.udev#L7 .

JakubVanek commented:
Hmm, ty 3D printer pravidla ale mají `SUBSYSTEMS=="usb"`, takže tam to vypadá, že se pletu.

ppisa commented:
OK, I have commented out installation in Debian rules... We can discuss that later but I do not want to introduce unintentional and unnoticed security hole. Another problem is to add user to tty group, because it allows to access terminal/even local of other users so probably to spy their activity and password entrance. dialout should be limited to serial ports and modems where modem daemon when link is used for serious link connection probably locks the device to itself... So that is relatively safe to grant user dialout group...

JakubVanek commented:
I think that commit 3464a23e872c995bc06a989c6dd2f33cf02bbab0 is still not effective - `dh_installudev` is called in the background, it just now installs the rule with a different priority. I have created https://github.com/cvut-fel-sdi/zero_elabviewer_community/pull/20 that contains the revert. Feel free to merge it into both this branch and the community master branch.

ppisa commented:
OK, I am for keeping logic but only disable it to not be active by default. But we can left that for another round now. I need to go to Elekroline.cz to discuss [silicon-heaven](https://github.com/silicon-heaven) its integration with [NXboot](https://nuttx.apache.org/docs/latest/applications/boot/nxboot/index.html) and [pysimCoder](https://github.com/robertobucher/pysimCoder) which are much larger and important projects, without unresponsive maintainers and with deadline for GSoC approaching...

JakubVanek commented:
ACK

adamberlinger commented:
Thank you very much for this work and sorry for a bit later response. I really appreciate this initiative.
I tested building the application in my Windows environment I noticed 2 small issues, or maybe it is just my mistake:
1. In `data/resources.rc` I had to put prefix `../data/` to compile properly otherwise I got following error message:
    ```windres.exe: can't open icon file `icons/baked/zero-elabviewer.ico': Invalid argument```
2. When I compile with `cmake` I need to pass it the path to my compiler (which is expected). However I had to specify the 2 additional compiler flags `-mwindows` and `-std=c++11`. The first one prevents pop-up of the console window and second is required for successful compilation. Maybe these flags should be added to the `CMakeLists.txt`?
I'm also using quite old GCC compiler v4.9.2 shipped with QT v5.6.3, so maybe part of the issue comes from this.

I'm also thinking if it would make sense to squash the commits, since there are some back and forth changes, e.g. regarding the name of the application. I understand the history behind, but maybe we can simplify here.
On the other hand, I don't want to bother you too much with some additional work.

JakubVanek committed:
fix: Enable C++11 by default

This was previously specified in the qmake packaging and was not carried
into the cmake packaging.

JakubVanek committed:
fix: Build application on Windows as a Win32 GUI application

This should hide the console window

JakubVanek committed:
fix: Copy Windows resources and icon to binary dir before build

This should fix an issue where resources.rc doesn't work with
a relative path inside the file.

JakubVanek commented:
Hello,

thank you! I tried to fix these issues in https://github.com/ppisa/zero_elabviewer/pull/1 (the PR is to the branch that this PR is from). I did not test building on Windows yet, but it sounds like the code did not break too much.

JV

JakubVanek commented:
> I'm also thinking if it would make sense to squash the commits, since there are some back and forth changes, e.g. regarding the name of the application. I understand the history behind, but maybe we can simplify here.

I think that squashing would indeed be OK here. While I usually strive for more granular history, the number of commits here is quite high and fixing up the history commit-by-commit could be quite difficult.

ppisa commented:
I agree that number of commits and finding the path and overcoming the stumbles makes history convoluted. But I would suggest to keep at least some commit separate, i.e. conversion to CMake, fixes, adding MAC OS etc.... I am quite busy but I hope to look at it till end of the week if nobody else has time and some skills/experience missing.
As for future and longer term maintainability and generally highlighting doc. Fischer's SDI decades of effort, I would suggest to move project into SDI group.
As for windows and `-mwindows and -std=c++11`, it reminds me same ancient time, but I have compiled Windows binaries only for my users, because I have never used any Windows after 3.1 on my desktop and even 3.1 only occasionally at its time. I do not see the `-mwindows` in any of my, Elektorline.cz CMake based projects. In my repo directory with lot of projects, I can locate about 4590 CMakefiles and  I see the option only in some local copy of CMake (its build for REACT OS), it seem to be in CMake mainline

https://github.com/Kitware/CMake/blob/master/Modules/Platform/Windows-GNU.cmake#L137

https://github.com/Kitware/CMake/blob/master/Modules/Platform/CYGWIN-GNU.cmake#L27

and in FlTK project

https://github.com/fltk/fltk/blob/master/CMake/variables.cmake#L81

I am not sure what is actual practice, @JakubVanek and @jirimaier could help there.

ppisa commented:
As for the long term SDI sustainability, I have negotiated that "contractors" of the project. doc. Fischer and doc. Petrucha has joined the https://github.com/orgs/cvut-fel-sdi project (@fischerjan321 and @petruvoj) so if prefer to keep mainline/root the project under your personal account, then please add them as cooner as the long term backup. But I would suggest the move. I have massively used projects under CVUT group to allow them long future ([QtRvSim](https://github.com/cvut/qtrvsim), [QtMips](https://github.com/cvut/QtMips/)).

JakubVanek commented:
> I am not sure what is actual practice, @JakubVanek and @jirimaier could help there.

Hi, I tried to find common practise on the web and created https://github.com/ppisa/zero_elabviewer/pull/1 with the fixes.
* `set_target_properties(zero_elabviewer PROPERTIES WIN32_EXECUTABLE TRUE)` instead of `-mwindows` (sourced from [here](https://stackoverflow.com/a/54429317)). DataPlotter seems to do the same thing.
* `set(CMAKE_CXX_STANDARD 11)` for C++11
* the remaining commit tries to fix the Windows icon/resources issue by moving the `.rc` and `.ico` files next to each other (inspired by DataPlotter).

ppisa commented:
Thanks, I have integrated updates/fixes to my pull branch. The series is huge, it needs reorganization definitely, it is unfortunate, that it has not been integrated i some more logical steps during last year...

JakubVanek commented:
I have created a branch that tries to make the history cleaner: https://github.com/cvut-fel-sdi/zero_elabviewer_community/tree/cleaned-up-history

I have made some additional code changes:
* I have changed the publishing organization name from "SDI Hackers" to "FEE CTU"
* I have removed the `<supports>` USB VID:PID from the AppStream metadata that I originally added. The VID:PID was intended to cover the F0-Lab, but the software is usable with serial port devices in general.

adamberlinger commented:
Thank you. I tested the update and it fixes the 2 point with my Windows build. I think that moving the project to some shared repository like https://github.com/cvut-fel-sdi/zero_elabviewer_community would be better and I can keep this repositor as a fork of the project.

ppisa commented:
Thanks for the report. As for the repository moving, there are two options, if you move repo and do not create new one in the old place, then GitHub should provide redirect forever from old repo to new one and you can use the SDI one as your main repo. Other option is to fork main repo back under your home and then you can use it for experiments and send finished ones through push or merges to the main repo. In each case, there is ensured the clean hierarchy where is main repository to be updated and used for pull requests.

You can proceed with move, try to consolidate history or wait some time till I get to it. But I have to focus on my subjects and projects at least till end of this week.

JakubVanek commented:
I have created https://github.com/ppisa/zero_elabviewer/pull/2 which adds an automated Windows build using a setup (hopefully) similar to the manual one. The resulting zipfile can be downloaded here: https://github.com/ppisa/zero_elabviewer/actions/runs/14208567876/artifacts/2863673762 . You can find it like this:
1. locate a commit
2. click on the tick mark
3. click on Details on the "WinXP build" row
4. open the "Run actions/upload-artifact@v4" spoiler
5. the URL should be at the end of the spoiler

The changes should apply cleanly also on top of the branch with the cleaned history at https://github.com/cvut-fel-sdi/zero_elabviewer_community/tree/cleaned-up-history .

EDIT: I have also added a build configuration that uses a more modern setup (Qt 5.15, MSVC 2019, FFTW 3.3.10). With this finished, I'll probably retreat for some time.

adamberlinger commented:
I will try to do the rebase and merge some commits by the end of this week and upload it here. Maybe first as a separate branch, just to double check if I haven't messed up anything.

adamberlinger commented:
I did some work on the rebase and at the moment I uploaded it to separate branch: [linux-pkg-rebase](https://github.com/adamberlinger/zero_elabviewer/tree/linux-pkg-rebase). There is still some work to be done, to remove some back and forth changes and make it more consistent.

Regarding the move to [cvut-fel-sdi](https://github.com/cvut-fel-sdi/), I think I can move the repository from my account here. And later we can remove the "_community" repository, after all the relevant data was moved to the main repository.
Or do you think there is better way to do this?

JakubVanek commented:
> Regarding the move to cvut-fel-sdi, I think I can move the repository from my account here. And later we can remove the "_community" repository, after all the relevant data was moved to the main repository.

From my POV this would be the ideal solution. However, before deleting the `zero_elabviewer_community` repository, I would like to ask how the new repository should/will be managed wrt. future contributions (including potentially mine):
- How many maintainers must approve of a pull request for it to be merged? (i.e. if one is enough or some kind of majority vote is preferred, and who can/should vote in that case)
- Similarly, if one of the maintainers posts a pull request, is an ACK from another maintainer required? If so, would there be a time window (a month?) after which the original poster could merge the PR by themselves if they received no feedback?

A good example for this might be https://github.com/cvut-fel-sdi/zero_elabviewer_community/pull/9 . In the community repo, I could merge my pull request without excessive guilt. However, in the new official repo, I think me merging it myself might be a bit too much. On the other hand, I would like to mitigate the scenario behind https://github.com/adamberlinger/zero_elabviewer/pull/4 where there was no clear path forward for a while.

@adamberlinger @ppisa

JakubVanek commented:
I propose the following policy:

- Only one maintainer ACK is required for merging a third-party contribution.
- Changes by a maintainer need to be ACKed by at least one other maintainer. However, if there is no feedback within one month from the time of the last push, the maintainer is free to merge their change.
- The necessity to have the code reviewed does not apply to the original program author.
  - This should reduce friction during development.
- Any maintainer can NACK a change and this overrides any previous ACKs.
  - This is intended to define how the (not publicly visible) argument around https://github.com/cvut-fel-sdi/zero_elabviewer_community/pull/14 should be handled - I shouldn't have merged the change, as @ppisa expressed his disagreement with one aspect of the change (udev rules installed by default) via a different channel.
  - The NACK should be used sporadically - only when the change would have a non-negligible negative impact.
- For Zero eLab Viewer, maintainers would be the members of @cvut-fel-sdi and the original program author would be @adamberlinger 

Would you be OK with this proposal?

adamberlinger commented:
Yes, I agree with the guidelines you proposed. I will also try to be more responsive to the incoming contributions.

adamberlinger commented:
I finalized the rebasing on the separate branch: [linux-pkg-rebase](https://github.com/adamberlinger/zero_elabviewer/tree/linux-pkg-rebase).
Please let me know if you have some remark, or you think there is something wrong. I modified some of the commit messages to match with the modified content. I tried to remove the "-community" naming, since it is removed later in the history.

If this is ok, I will merge it to the main branch and move it to @cvut-fel-sdi

JakubVanek commented:
I went quickly through it and everything seems OK to me. Thank you!

ppisa commented:
Thanks for the effort, I have run local packaging and DEB build successfully and pushed builds to SUSE Open Build Service as well as Ubuntu Launchpad

Suse finished successfully (except openSUSE_Leap_15.2_ARM which is old and uninteresting)
https://build.opensuse.org/package/show/home:ppisa/zero-elabviewer

Ubuntu Launchapd 
https://code.launchpad.net/~cvut-fel-sdi/+recipe/zero-elabviewer-build

JakubVanek commented:
Thank you all!

ppisa commented:
Some remark to Ubuntu Launchpad packaging.

The repository sync to Launchpad is setup on Launchpad mirror

https://code.launchpad.net/~cvut-fel-sdi/zero-elabviewer/+git/zero-elabviewer

It points to https://github.com/ppisa/zero_elabviewer now.

As the packaging is included in mainline now, the repository base should be updated. I am waiting with it if you decide to move repository to https://github.com/orgs/cvut-fel-sdi/repositories group or not.

Then I try to update Lauchpad to run the build from the official repository.

I have tried to add you @adamberlinger to https://launchpad.net/~cvut-fel-sdi/ Lauchpad team, if you confirm, that it is really you, I elevate your role to team administrators.

There is some hack used to deliver preprocessed Debian control file to Launchpad. I keep that file in preprocessed-for-packaging branch

  https://github.com/ppisa/zero_elabviewer/blob/preprocessed-for-packaging/debian/control

If you agree, I would put that single file orphan branch to the final repo to allow Launchpad packaging from single repo. Other option is to deliver Debian control file with all fields filled in some directory of the master branch. But I have some personal preference to keep no generated files in the main source tree. So this trick with separate branch which in the case QtRvSim has not been updated through years seem to me as nicer.

Best wishes,

Pavel Pisa
